tests/unit/test_retry.py

```diff
@@ -467,3 +490,425 @@ def test___init___when_retry_is_executed(self, sleep, uniform):
         assert _some_function.call_count == 2
         target.assert_has_calls([mock.call("meep"), mock.call("meep")])
         sleep.assert_any_call(retry_._initial)
+
+    def _generator_mock(
+        self,
+        num=5,
+        error_on=None,
+        return_val=None,
+        exceptions_seen=None,
+        ignore_sent=False,
+    ):
+        """
+        Helper to create a mock generator that yields a number of values
+        Generator can optionally raise an exception on a specific iteration
+        """
+        try:
+            sent_in = None
+            for i in range(num):
+                if error_on and i == error_on:
+                    raise ValueError("generator mock error")
+                sent_in = yield (sent_in if sent_in else i)
+                if ignore_sent:
+                    sent_in = None
+            return return_val
+        except (Exception, BaseException, GeneratorExit) as e:
+            # keep track of exceptions seen by generator
+            if exceptions_seen is not None:
```

Comment:
Actually I had to revert this; we want to modify the list that was passed in instead of replacing the list here.

I added some documentation around this function